#Conjunto de dados
- **AT**(Ambient Temperature): Temperatura em °C. Afeta a densidade do ar e a eficiência da combustão

- **V**(Exhaust Vacuum): Vácuo de escape em cm Hg. Relacionado ` a pressão na turbina a vapor.

- **AP**(Ambient Pressure): Pressão atmosférica em mbar.

- **RH**(Relative Humidity): Umidade relativa em porcentagem.

- **PE**(Net Hourly Electrical Energy Output): A variável alvo em MW (MegaWatts).

#1 - Calcule a correlação entre as variáveis independentes (AT, V, AP, RH) e a variável dependente (PE). Quais variáveis parecem ter maior influência na produção de energia?

In [ ]:
import pandas as pd

df = pd.read_excel('UCI CCPP.xlsx')

print(df.head())

      AT      V       AP     RH      PE
0  14.96  41.76  1024.07  73.17  463.26
1  25.18  62.96  1020.04  59.08  444.37
2   5.11  39.40  1012.16  92.14  488.56
3  20.86  57.32  1010.24  76.64  446.48
4  10.82  37.50  1009.23  96.62  473.90


In [9]:
#Removendo a coluna PE do dataframe x
x = df.drop(columns=['PE'])
#data frame da coluna excluida (y)
y = df['PE']

#Correlação geral
corelacao = df.corr()
print(corelacao)

          AT         V        AP        RH        PE
AT  1.000000  0.844107 -0.507549 -0.542535 -0.948128
V   0.844107  1.000000 -0.413502 -0.312187 -0.869780
AP -0.507549 -0.413502  1.000000  0.099574  0.518429
RH -0.542535 -0.312187  0.099574  1.000000  0.389794
PE -0.948128 -0.869780  0.518429  0.389794  1.000000


In [11]:
#correlação entre x e y
corelacao_x_y = x.corrwith(y) #calcula a correlação entre o df x e y
print(corelacao_x_y)

AT   -0.948128
V    -0.869780
AP    0.518429
RH    0.389794
dtype: float64


A variável que apresenta maior influência na produção de energia é a temperatura ambiente (AT), pois possui a maior correlação em valor absoluto com a variável dependente (PE), indicando forte relação inversa. Isso significa que, à medida que a temperatura aumenta, a produção de energia tende a diminuir

Ranking de influência de valor absoluto
|Tipo|Grau|
|:---|:---|
|AT|Top 1|
|V|Top 2|
|AP|Top 3|
|RH|Top 4|

#2 - Calcule a correlação entre as variáveis independentes (AT, V, AP, RH) entre si. Existe alguma relação forte entre elas? Se sim, isso pode indicar multicolinearidade.

In [13]:
import pandas as pd

df = pd.read_excel('UCI CCPP.xlsx')

print(df.head())

      AT      V       AP     RH      PE
0  14.96  41.76  1024.07  73.17  463.26
1  25.18  62.96  1020.04  59.08  444.37
2   5.11  39.40  1012.16  92.14  488.56
3  20.86  57.32  1010.24  76.64  446.48
4  10.82  37.50  1009.23  96.62  473.90


In [17]:
#correlação entre se geral
x = df.drop(columns=['PE'])

correlacao_x = x.corr()
print(correlacao_x)

          AT         V        AP        RH
AT  1.000000  0.844107 -0.507549 -0.542535
V   0.844107  1.000000 -0.413502 -0.312187
AP -0.507549 -0.413502  1.000000  0.099574
RH -0.542535 -0.312187  0.099574  1.000000


Regra prática
|Correlação|Interpretação|
|:---|:---|
|>0,8 ou < -0,8|Forte|
|0,5 a 0,8|Moderada|
|<0,5|Fraca|

Ao analisar a correlação entre as variáveis independentes, observa-se que as variáveis AT e V apresentam uma correlação forte entre si. Isso pode indicar a presença de multicolinearidade, o que pode afetar a estabilidade dos coeficientes no modelo de regressão linear. As demais variáveis apresentam correlação fraca a moderada

#3 - Construa um modelo de regressão linear múltipla utilizando as variáveis independentes (AT, V, AP, RH) para prever a variável dependente (PE). Divida os dados em treino (80%) e teste (20%).

#4 - Avalie o desempenho do modelo utilizando métricas como R², RMSE e MAE. O modelo é adequado para prever a produção de energia? Justifique sua resposta.

In [5]:
import pandas as pd

df = pd.read_excel('UCI CCPP.xlsx')

print(df.head())

      AT      V       AP     RH      PE
0  14.96  41.76  1024.07  73.17  463.26
1  25.18  62.96  1020.04  59.08  444.37
2   5.11  39.40  1012.16  92.14  488.56
3  20.86  57.32  1010.24  76.64  446.48
4  10.82  37.50  1009.23  96.62  473.90


Como podemos notas temos mais de 2 variaveis, ou seja usaremos a regressão linear múltipla

$$
y = \theta_0 + \theta_1 x_1 + \theta_2 x_2 + ... + \theta_px_p + \epsilon
$$

em que:

$\bullet$ $x_1, x_2, ... , x_p$ são variáveis independentes (Tambem chamdas de características ou features);

$\bullet$ $y$ é a variável dependente (valor no qual desejamos aproximar ou prever);

$\bullet$ $\theta_0$ é o coeficiente linear da reta;

$\bullet$ $\theta_1, \theta_2, ... , \theta_p$ são os coeficientes que representam o impacto de cada variável independente em $y$

$\bullet$ $\epsilon$ é o erro (ou resíduo) da aproximação


Usamos todas as variaveis para esse cálculo por tanto nosso `x` e `y` é dado por:

In [4]:
#Removendo a coluna PE do dataframe x
x = df.drop(columns=['PE'])
#data frame da coluna excluida (y)
y = df['PE']

Como podemos notar, temos uma grandeza de tamanhos bem diferente entre nossas variáveis, com isso temos que usar uma normalização `(Padronição z-score, Normalização (Min-Max scaling))`
  
- Padronização z-score = $\frac{x - 𝝁}{\sigma}$
- Normalização (Min-Max scaling) = $x_{norm} = \frac{x - x_{min}}{x_{max}- x_{min}}$

Vamos analisar os 2


In [13]:
#Padronização z - score
from sklearn.preprocessing import StandardScaler
z_score = StandardScaler()
#Ajustar e transformar os dados
x_z_score = z_score.fit_transform(x)  #fit() = calcula: média (μ) e desvio padrão (σ) e transform() aplica a formula

print(x_z_score)

[[-0.62951938 -0.98729659  1.8204884  -0.00951935]
 [ 0.74190911  0.68104512  1.1418628  -0.97462052]
 [-1.95129733 -1.17301765 -0.18507756  1.2898397 ]
 ...
 [ 1.56583973  1.57581141 -0.05709854 -2.52261812]
 [ 0.64797565  1.19177803  0.1011913  -0.74790051]
 [ 0.26150656  0.64641916  0.66867722 -0.37254534]]


In [17]:
#Normalização (Min-Max scaling)
from sklearn.preprocessing import MinMaxScaler
min_max = MinMaxScaler()
#Ajustar e transformar os dados
x_min_max = min_max.fit_transform(x)  #fit() = encontra: valor mínimo (xmin) e valor máximo (xmax) transform() aplcia na formula

print(x_min_max)

print(x_min_max.min())
print(x_min_max.max())

[[0.37252125 0.29181495 0.77159119 0.63820375]
 [0.66203966 0.66903915 0.6718634  0.44932976]
 [0.09348442 0.24982206 0.47686216 0.8924933 ]
 ...
 [0.83597734 0.87135231 0.49566939 0.1463807 ]
 [0.64220963 0.78451957 0.51893096 0.49369973]
 [0.56062323 0.66120996 0.60232616 0.56715818]]
0.0
1.0000000000000002


Resultados:

z-score = $[-\infty, \infty]$

Min-Max-sacaling = [0, 1]

Melhores ocasiões para o uso de cada um dela

|Critérios|Z-score|Min-Max-Scaling|
|:---|:---|:---|
|Outliers|Lida Melhor|Sensível|
|Distribuição|Normal(Ideal)|Qualquer|
|Interpretação|Média = 0|Escala fixa|
|Uso comum|Estatística/ML|Deep Learning|

O Z-score é mais adequado quando há presença de outliers e quando se deseja padronizar os dados com média zero e desvio padrão um. Já o Min-Max Scaling é utilizado quando se deseja limitar os dados a um intervalo específico, geralmente entre 0 e 1.

Nesse momento vamos dividir as amostras em treino e teste

In [20]:
from sklearn.model_selection import train_test_split

#Dividindo os dados em treino e teste

x_train, x_test, y_train, y_test = train_test_split(
    x, y,
    test_size=0.2, # 20% dos dados são testes e 80% são treinos
    random_state=42 #Garante que a divisão seja sempre igual
)

print(f"Tamanho do datasheet: {len(df)}")
print(f"Tamanho do dataset de treino: {len(x_train)}")
print(f"Tamanho do dataset de teste: {len(x_test)}")


Tamanho do datasheet: 9568
Tamanho do dataset de treino: 7654
Tamanho do dataset de teste: 1914


Por fim vamos usar o modelo de regressão linear disponivel na `sklearn`

In [42]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x_train, y_train)

#model.score = qualidade do modelo
print("R2 treino = ",model.score(x_train, y_train))
print("R2 teste = ",model.score(x_test, y_test))

R2 treino =  0.928331545565795
R2 teste =  0.9301046431962188


interpretação prática do $R^2$
|Situação|Diagnóstico|
|:---|:---|                                      
|Treino alto, teste baixo|Overfitting|
|Treino $\approx$ Teste|Modelo bom|
|Ambas baixa|Modelo ruim|

Satisfação
|R2|>|0,8|Modelo muito bom|
|:---|:---|:---|:---|
|R2|Entre|0,6 e 0,8|Razoavel|
|R2|<|0,5|Fraco|

Determinando os coefs $\theta$

In [34]:
theta = model.coef_
print(theta)

[-1.98589969 -0.23209358  0.06219991 -0.15811779]


$\hat{y} = -1.98589969 - 0.23209358x_1 + 0.06219991x_2 - 0.15811779x_3$

Formula
$$
R^2 = 1 - \frac{SQR}{SQT} → 1 - \frac{\sum^{N_t - 1}_{i=0}(y^{(i)}- \bar{y})}{\sum^{N_t - 1}_{i=0} (y^{(i)}- \hat{y}^{(i)})^2}
$$

- **SQR:** soma dos quadrados dos resíduos: variação que o modelo não conseguio explicar(Diferença entre os valores reais e os valores preditos)
- **SQT:** Soma total dos quadrados, representa a variação total dos dados em relação à média

Erros quadratico médio(MSE)
$$
MSE = 1 - \frac{1}{N_T}\sum^{N_t-1}_{i=0}(y^{(i)}-\hat{y}^{(i)})^2
$$

$$
RMSE = \sqrt{MSE}
$$


In [41]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

y_pred = model.predict(x_test)

mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)


print(f"Coeficiente de determinação R2 = {r2:.4f}")
print(f"Erro Quadrático Médio (MSE) = {mse:.4f}")
print(f"Raiz do erro quadratico médio = {rmse:.4f}")
print(f"Erro Absoluto Médio (MAE) = {mae:.4f}")

Coeficiente de determinação R2 = 0.9301
Erro Quadrático Médio (MSE) = 20.2737
Raiz do erro quadratico médio = 4.5026
Erro Absoluto Médio (MAE) = 3.5959


Com esses resultados e pela tabela de satisfaação acima, podemos dizer que é um bom modelo para essa análise.